### ISSUE TITLE
Security vulnerability disclosure CVE-2024-0864

### ISSUE THREAD

**Issue by kazet** [leokhoa/laragon#613]
Hello,

CERT PL found a security vulnerability in this repository. How can we report this privately? We don't see any security policy describing how such vulnerabilities should be reported.

**Comment by rdggithub** [comment]
Was this solved somehow?

**Comment by TheOldMan2000** [comment]
This report should have been treated a lot better.

**Comment by alxndr-w** [comment]
@kazet, if this is still valid, check out this PR https://github.com/leokhoa/laragon/pull/988 - also, if it's still valid, you can contact me and I will give it to @leokhoa directly.

@TheOldMan2000 you're right. 

Your comment was in early 2024 and maybe you are happy to hear that Laragon is now in active support again and I'm optimistic that things like these will be addressed faster (or at all).

**Comment by rdggithub** [comment]
FYI: The mentioned security issue was/is a remote code execution, details [here](https://www.cve.org/CVERecord?id=CVE-2024-0864).

I find it really disturbing how this was/is treated down to the present day.

**Comment by alxndr-w** [comment]
> I find it really disturbing how this was/is treated down to the present day.

Please have a look at the pinned discussions which will explain why and plans to actively support Laragon again from now on.

This can't happen over night unfortunately.

GitHub for example offers great security services and I will strongly suggest to start using them. I will keep an eye on that.

**Comment by Sophist-UK** [comment]
I am with @rdggithub here - Leo's priorities are completely wrong. 

He has been focused on pricing for well over a month now, and still hasn't outlined what his plans are (if ANY????) for enhancing v7 to justify the licensing costs, but he cannot deal with it for over 14 months. He apparently thinks that allowing the community some input is somehow a security issue - YET when it comes to a genuine security issue he simply ignores it.

@alxndr-w 's response to this is equally bad. Any professional software developer would provide security-related support for old versions - Microsoft is STILL doing this for Windows 10 despite the number of years that Windows 11 has been the current version. Any professional software developer would A) acknowledge the security issue, B) issue regular updates explaining the seriousness from the developers perspective (because they understand the code best) and C) explain when and how they fixed it (or intend to fix it). We have no idea whether this security issue persists in v7 and dealing with it by posting a disclaimer is a disgrace and yet another example of unprofessionalism. 

If @leokhoa wants to charge money for Laragon v7 he pretty much had better explain to people why they should pay it, and thus far he is doing a non-existent job of doing this, and instead is demonstrating again and again just how amateur he is being. 

Unfortunately, I am not optimistic that things are going to change for the better. Leo has stated elsewhere here that he is using 15yo development tools and has no plans to change. He has told me that he doesn't want community input, doesn't want to develop or publicise any sort of vision for how Laragon is going to be enhanced, and may in fact become more secretive (rather than more open) because of his unjustified concerns about security vulnerabilities. 

TBH I just cannot see why people will pay a significant amount for a commercial license (not on the basis of actually delivered better functionality, nor on the basis of future enhancements that have been visualised and promised, but rather simply because Leo wants to make money from us. 

I was never going to be a customer for a commercial license - only for a non-commercial one - but from my own perspective given the amateurism I am seeing again and again, I have been UN-SOLD on buying a v7 license, and I would rather use a free v6 than a chargeable v7 regardless of how low the price for a non-commercial license may go.

**Comment by alxndr-w** [comment]
> There was enough time to build and promote a paid license around the former "forever free" model, but no time for a SA or warning for its users?

I don't justify how this was handled, I wrote more context why this happened. You are asking to get this right. I ask to get your expectations right. Because time machines aren't invented yet. I bet, @leokhoa was not aware of this. Which is bad, but can't be undone.

Today, I tried to reproduce the issue, but:

1. I wasn't able to check out where this upload script is used - I still can't find it in use.
2. I wasn't able to find the original source to this code and if there's an updated version since it mentions Internet Explorer 7.
3. And, the CVE document <https://www.cve.org/CVERecord?id=CVE-2024-0864> does not explain enough how this could be used in the wild. So it's really difficult to take this CVE serious. Maybe it's because of another project, where people added CVEs that are clearly no bugs or security issues at all, why I first want to understand where and how this is a flaw. 
4. IMHO, if this was a feature to upload files via browser in Laragon, then it would be the best to delete this feature at all. It sounds like a very bad idea in general that once might was a good intention.
5. Compared to have components of Laragon not updated for a long time, this vulnerability is ONE problem that needs to be addressed. Further more, handling things like thes properly in general is another issue.
6. Microsoft isn't the best example to neither in the past nor now to handle vulnerabilities.

I don't feel responsible for fixing the issue, so attacking me is just unnecessary drama.

Laragon needs to be secure. There's no doubt about that.

So please let me know how to recreate this issue - send me an email if possible.

**Comment by leokhoa** [comment]
@alxndr-w 
I can confirm that this issue was resolved in Laragon 7.0.0, released on December 16, 2025.
To clarify, **the code in question is PHP code from an open-source project for sharing data and is not part of the Laragon engine itself.**
This issue has reminded us to exercise **GREATER CAUTION when adding script code** to Laragon ecosystem.
Thank you for creating the Security.md file - it's a good step toward enhancing Laragon's security moving forward.

**Comment by Sophist-UK** [comment]
@alxndr-w 

If @leokhoa wasn't aware of someone reporting a security issue via Github, then shame on him. But at the time it was reported it was a free product and so support was definitely on an ad-hoc basis.

HOWEVER, now it isn't - Leo is expecting people to pay for Laragon and he is expecting commercial developers to pay quite a lot. And if he wants them to do that he has got to be a heck of a lot more professional about the whole thing than he is being at the moment.

It doesn't really matter who decided to deal with the v vulnerability with a disclaimer and without (until a minute ago having any further details) because it is the Laragon Project that needs to be professional and NOT deal with it in an rank amateur fashion.

Now to Leo's more recent comment: "This issue has reminded us to exercise GREATER CAUTION when adding script code to Laragon ecosystem."

Yes - there needs to be some caution in selecting products for inclusion, but it is unclear whether this product https://github.com/LPology/Simple-Ajax-Uploader was explicitly included by Leo or was simply part of another product. Certainly choosing minor products rather than ones from established brands with good security recommendations can sometimes be risky - but often you make sensible choices and yet a CVE is issued anyway.

You cannot protect yourself completely from having security issues - the real question is how they are dealt with, and IMO this one has been until now dealt with in a completely unprofessional way - but more importantly it is still being dealt with unprofessionally because:

* Leo is NOT attempting event to warn existing v6 users
* Leo is not contemplating fixing this in v6.

And the SECURITY.md file is IMO a shockingly bad approach - it is NOT a good step towards enhancing Laragon's security moving forwards because you do NOT deal with Security issues in this way.

**Comment by rdggithub** [comment]
> I don't justify how this was handled, I wrote more context why this happened. You are asking to get this right. I ask to get your expectations right. Because time machines aren't invented yet. I bet, [@leokhoa](https://github.com/leokhoa) was not aware of this. Which is bad, but can't be undone.

Personally I stopped using laragon in early 2024 when this issue wasn't addressed in any way. My expectations though are quite simple: create an SA/security notice on the security page publishing all the details you have. You would be free to add your thoughts on how this is mitigated, or not a valid CVE in your opinion. And could classify it as low impact, if you think so.

> 
> Today, I tried to reproduce the issue, but:
> 
>     1. I wasn't able to check out where this upload script is used - I still can't find it in use.
> 
>     2. I wasn't able to find the original source to this code and if there's an updated version since it mentions Internet Explorer 7.

The original code should be this one? https://github.com/LPology/Simple-Ajax-Uploader

>     3. And, the CVE document https://www.cve.org/CVERecord?id=CVE-2024-0864 does not explain enough how this could be used in the wild. So it's really difficult to take this CVE serious. Maybe it's because of another project, where people added CVEs that are clearly no bugs or security issues at all, why I first want to understand where and how this is a flaw.

From what I understood after digging further, it's just that the tool is present in the laragon source, be it as an example or whatever. I'm sure @leokhoa could clarify things in a second, but since he seems to don't care beside promoting that this doesn't affect 7.x, I will not invest any more time here.

> I don't feel responsible for fixing the issue, so attacking me is just unnecessary drama.

I think no one attacked you here, at least not me. I'm not even aware of how exactly you are involved in the project, I just care for the users who aren't that alert or don't have the time to check issues / third party notifications.

**Comment by Sophist-UK** [comment]
I believe that this summary from @rdggithub has described my own concerns better than I could have done.

The sad thing about this is that @leokhoa has a whole bunch of people who want to help him succeed - to help with documentation, private and public beta testing, confidential code reviews, advice on communications, advice on how to handle security issues, documentation, UX redesign, product vision, maintaining up to date versions of services, extending the product functionality in tens of directions simultaneously, documentation, ... and Leo has neither the knowledge/prior experience to manage all of this nor the time to do it all.

In my opinion, if Leo delegated all the draft work for all of the above to the community and focused only on deciding whether to approve them and doing the coding for the executable and any other stuff that genuinely only he can do, then he would still have more work to do than he would have time for - so how he can possibly hope to deliver a professional service without relying on this help I have no idea.

The bottom line is that if Leo wasn't so stubborn and took advantage of this help rather than reject it, I think this could be a rip-roaring success rather than senior users such as @rdggithub and myself needing to criticise and then to walk away.

**Comment by alxndr-w** [comment]
<https://github.com/leokhoa/laragon/discussions/997>

Duy, a "former colleague of [Leo], an SRE and Cloud Expert based in New Zealand, holds a PhD in Network and Security" was announced to join the team.

As I already said, things need time but you can't assume that things will not change or not be addressed just because they have not in the past or not *yet*.

**Comment by leokhoa** [comment]
Laragon is great for local development, and the issue reporting doesn't happen in a local environment. Until now, Laragon is not recommended to run on public internet. Because:

1. MySQL doesn't have a password set by default.
2. It comes with some tools that are helpful for developers but might have security concerns (phpRedisAdmin, phpMyAdmin,...)
3. It allows older PHP versions that may have known security issues (CVE).


Moving forwards, if you find a security vulnerability in Laragon, please report it via email to Leo Khoa at [leo@laragon.org](mailto:leo@laragon.org)
We take security seriously, and all reports will be promptly addressed.
